tutorial/openmp.md

```diff
@@ -0,0 +1,573 @@
+
+```python
+#$ omp taskloop [clause[ [,]clause] ... ]
+for-loops
+```
+
+#### Syntax Atomic
+
+```python
+#$ omp atomic [clause[ [,]clause] ... ]
+structured-block
+#$ omp end atomic
+```
+
+#### Example
+The ``` #$ omp taskloop ``` construct specifies that the iterations of one or more associated loops will be executed in parallel using explicit tasks.\
+The ``` #$ omp atomic ``` is used to ensure that a specific storage location is accessed atomically.
+```python
+from pyccel.stdlib.internal.openmp import omp_get_thread_num
+
+x1 = 0
+x2 = 0
+#$ omp parallel shared(x1,x2) num_threads(2)
+
+#$ omp taskloop
+for i in range(0, 100):
+  #$ omp atomic
+  x1 = x1 + 1 #Will be executed (100 x 2) times.
+  #$ omp end atomic
+
+#$ omp single
+#$ omp taskloop
+for i in range(0, 100):
+  #$ omp atomic
+  x2 = x2 + 1 #Will be executed (100) times.
+  #$ omp end atomic
+#$ omp end single
+
+#$ omp end parallel
+print("x1 : ", x1);
+print("x2 : ", x2);
+```
+
+The output of this program is (you may get a different output, but the sum must be the same for each thread):
+```shell
+❯ pyccel omp_test.py --language c --openmp
+❯ ./omp_test
+x1 : 200
+x2 : 100
+```
+
+### Masked Construct
+
+#### Syntax
+
+```python
+#$ omp masked [ filter(integer-expression) ]
+structured-block
+#$ omp end masked
+```
+
+#### Example
+
+The ``` #$ omp masked ``` pragma is used here to specify a structured block that is executed by a subset of the threads of the current team.
+```python
+result = 0
+#$ omp parallel num_threads(4)
+#$ omp masked
+result = result + 1
+#$ omp end masked
+#$ omp end parallel
+print("result :", result)
+```
+
+The output of this program is:
+```shell
+❯ pyccel omp_test.py --language c --openmp
+❯ ./omp_test
+result : 1
+```
+
+### SIMD Construct
+
+#### Syntax
+
+```python
+#$ omp simd [clause[ [,]clause] ... ]
+loop-nest
+```
+
+#### Example
+
+The ``` #$ omp simd ``` pragma is used to transform the loop into a loop that will be executed concurrently using Single Instruction Multiple Data (SIMD) instructions.
+```python
+from numpy import zeros
+result = 0
+n = 1337
+arr = zeros(n, dtype=int)
+#$ omp parallel num_threads(4)
+#$ omp simd
+for i in range(0, n):
+  arr[i] = i
+#$ omp end parallel
+for i in range(0, n):
+  result = result + arr[i]
+print("Result:", result)
+```
+
+The output of this program is:
+```shell
+❯ pyccel omp_test.py --language c --openmp
+❯ ./omp_test
+Result: 893116
+```
+
+### Task / Taskwait Construct
+
+#### Syntax Task Construct
+
+```python
+#$ omp task [clause[ [,]clause] ... ]
+structured-block
+#$ omp end task
+```
+
+#### Syntax Taskwait Construct
+
+```python
+#$ omp taskwait
+```
+
+#### Example
+
+The ``` #$ omp task ``` pragma is used here to define an explicit task.\
+The ``` #$ omp taskwait ``` pragma is used here to specify a wait on the completion of child tasks of the current task.
+```python
+@types('int', results='int')
+def fib(n):
+  if n < 2:
+    return n
+  #$ omp task shared(i) firstprivate(n)
```

Comment:
the documentation didn't cover  the explanation of clauses

Reply:
Ah ok. This is probably just the one clause I didn't recognize or who's explanation didn't seem obvious from its name.

I'm not sure if we should add some explanation of clauses or if it is sufficient to describe the main commands and leave the user to find the clauses in openmp's documentation.

What do you think @yguclu , @saidctb ?

Reply:
My feeling is that we can leave the clauses for openmp to explain so we don't have to keep everything up to date. But it might be worth giving a link so the user knows where those definitions are, especially in cases where it might not be obvious from the name (e.g. do/for)

Reply:
i agree, i found an article from microsoft which describe the different clauses of openmp.
https://docs.microsoft.com/en-us/cpp/parallel/openmp/reference/openmp-clauses
in the openmp website there are just pdf files which contain both constructs and clauses.